### ISSUE TITLE
Command from a dependency not found

### ISSUE THREAD

**Issue by jluttine** [svanderburg/node2nix#68]
I'm trying to build Daedalus: https://github.com/input-output-hk/daedalus

I ran:
```
git clone https://github.com/input-output-hk/daedalus
cd daedalus
git checkout 0.8.0
nix-shell -p nodePackages.node2nix
node2nix
nix-build -A package
```

But I get the following error:
```
> react-polymorph@0.3.1 postinstall /nix/store/d2vcrlg6qph2x6k73q9k4wjsksh9kyss-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/react-polymorph
> postinstall-build lib --only-as-dependency

sh: postinstall-build: command not found

npm ERR! Linux 4.9.66
npm ERR! argv "/nix/store/rk14rjqz9b2wnwvr0mz8z5ww7frdm8db-nodejs-4.8.5/bin/node" "/nix/store/rk14rjqz9b2wnwvr0mz8z5ww7frdm8db-nodejs-4.8.5/bin/npm" "--registry" "http://www.example.com" "--nodedir=/nix/store/5ia5f7gi263xs5lb6zswpzhcabifnhxk-node-sources" "--production" "rebuild"
npm ERR! node v4.8.5
npm ERR! npm  v2.15.11
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! react-polymorph@0.3.1 postinstall: `postinstall-build lib --only-as-dependency`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the react-polymorph@0.3.1 postinstall script 'postinstall-build lib --only-as-dependency'.
npm ERR! This is most likely a problem with the react-polymorph package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     postinstall-build lib --only-as-dependency
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs react-polymorph
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls react-polymorph
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /nix/store/d2vcrlg6qph2x6k73q9k4wjsksh9kyss-node-daedalus-0.8.0/lib/node_modules/daedalus/npm-debug.log
builder for ‘/nix/store/6skhzcgf6qisawvhnqi6lfpxrs9sz55x-node-daedalus-0.8.0.drv’ failed with exit code 1
error: build of ‘/nix/store/6skhzcgf6qisawvhnqi6lfpxrs9sz55x-node-daedalus-0.8.0.drv’ failed
```
Any ideas what's wrong and how to fix or work around?

**Comment by svanderburg** [comment]
I haven't checked all details, but as far as I can see, this package depends on a script that is provided by the [postinstall-build](https://www.npmjs.com/package/postinstall-build) package. Yet I don't see it declared as a dependency or devDependency in the `package.json` configuration file.

My guess is that in an "ordinary" NPM-workflow this package is installed globally. You can simulate that kind of behaviour by defining a supplement JSON file containing this package and providing this JSON file as an additional parameter to `node2nix`. Then in the generated Nix builder environment, the supplement packages are provided as additional build inputs.

See the README.md section 'Adding additional/global NPM packages to a packaging process' for more information.

**Comment by jluttine** [comment]
It is provided as a dependency in `package.json` of  `react-polymorph` and appears in the resulting `node-packages.nix`. It's not a direct dependency of Daedalus which I'm trying to build, yet I thought it should work.

I'll try that supplement JSON file at some point, thanks.

**Comment by jluttine** [comment]
Adding it as a supplement solved that issue, thanks. However, now there's an issue with some other dependency:

```
> electron-chromedriver@1.6.0 install /nix/store/pdjzhy44b5m3g2f2hnd878gpsxilzlgq-node-dependencies-daedalus-0.8.0/daedalus/node_modules/spectron/node_modules/electron-chromedriver
> node ./download-chromedriver.js

/nix/store/pdjzhy44b5m3g2f2hnd878gpsxilzlgq-node-dependencies-daedalus-0.8.0/daedalus/node_modules/spectron/node_modules/electron-chromedriver/download-chromedriver.js:16
  if (error != null) throw error
                     ^

Error: getaddrinfo ENOTFOUND github.com github.com:443
    at errnoException (dns.js:27:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:78:26)

npm ERR! Linux 4.9.68
npm ERR! argv "/nix/store/rk14rjqz9b2wnwvr0mz8z5ww7frdm8db-nodejs-4.8.5/bin/node" "/nix/store/rk14rjqz9b2wnwvr0mz8z5ww7frdm8db-nodejs-4.8.5/bin/npm" "--registry" "http://www.example.com" "--nodedir=/nix/store/5ia5f7gi263xs5lb6zswpzhcabifnhxk-node-sources" "rebuild"
npm ERR! node v4.8.5
npm ERR! npm  v2.15.11
npm ERR! code ELIFECYCLE
npm ERR! electron-chromedriver@1.6.0 install: `node ./download-chromedriver.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the electron-chromedriver@1.6.0 install script 'node ./download-chromedriver.js'.
npm ERR! This is most likely a problem with the electron-chromedriver package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./download-chromedriver.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs electron-chromedriver
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls electron-chromedriver
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /nix/store/pdjzhy44b5m3g2f2hnd878gpsxilzlgq-node-dependencies-daedalus-0.8.0/daedalus/npm-debug.log
builder for ‘/nix/store/ralsq7qsfnb48hhh9fgj0vmnix1rylhp-node-dependencies-daedalus-0.8.0.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/cprkh4xa086nxpkanpnasy36w065hapc-daedalus-0.8.0.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/cprkh4xa086nxpkanpnasy36w065hapc-daedalus-0.8.0.drv’ failed
/run/current-system/sw/bin/nix-shell: failed to build all dependencies
```

I tried adding `electron-chromedriver` to supplement JSON file, but it didn't help. Also, I tried disabling running installation for that package with:
```
nodePackages // {
  # Disable npm install for electron-chromedriver because it would download
  # chromedriver
  electron-chromedriver = nodePackages.electron-chromedriver.override (oldAttrs: {
    dontNpmInstall = true;
  });
}
```
Didn't help. Any ideas?

**Comment by svanderburg** [comment]
This is a really annoying issue that is basically caused by an undesired property of some NPM packages. I have encountered several NPM packages that include a script that attempts to download artifacts from external locations, such as precompiled binaries. This solution is basically a limitation of NPM that some package authors try to bypass by implementing their own custom download scripts.

Since downloads in pure environments are restricted, the deployment will fail as a result.

The only way to "properly" fix this package is by patching it -- you can override the expression in such a way that Nix obtains the corresponding artifact (e.g. the binary that the `download-chromedriver.js` script attempts to download). Then add a `preRebuild` hook that patches the script (or disables the script directive in `package.json`) to prevent NPM from downloading it.

Unfortunately, there is no easy solution for this.

However, there is a package: `webdrvr` in the Nixpkgs set that requires a similar override.:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/default-v6.nix

In the above expression, we override the expression and we patch the package to prevent NPM downloading selenium. Instead, we obtain it ourselves.

Maybe you can use this strategy as an example to fix the electron chromedriver package.

**Comment by kreisys** [comment]
tl;dr, but it looks very similar to a problem I had with bcrypt not finding node-pre-gyp. I was banging my head against the wall for a few good hours but the workaround turned out to be pretty simple (ugly but simple): run `npm rebuild --ignore-scripts` prior to `npm rebuild`. If I remember correctly my conclusion was that strangely `npm rebuild` first tries to run the package scripts and only then creates the symlinks in `node_modules/.bin` so adding `--ignore-scripts` lets it reach the symlinking phase without dying first. I patched it in my fork. Looky here: https://github.com/kreisys/node2nix/commit/ed36f797cff793b95a0a7df83a86832ef18c9d36

Hope this helps!

**Comment by svanderburg** [comment]
@kreisys What you can also do is adding node-pre-gyp as a supplement package, e.g. by creating a supplement.json config and regenerating the expressions with node2nix. This is what I did to support newer bcrypt versions.

As far as I know, node-pre-gyp packages typically run a fallback script that will build the package from source if needed.

**Comment by onny** [comment]
> Adding it as a supplement solved that issue, thanks. However, now there's an issue with some other dependency:
Could you please share your solution? I'm facing a similar packaging issue
